crates/ra_cargo_watch/src/lib.rs

```diff
@@ -383,23 +368,6 @@ impl WatchThread {
         } else {
             None
         };
-        WatchThread { handle, message_recv }
-    }
-}
-
-impl std::ops::Drop for WatchThread {
-    fn drop(&mut self) {
-        if let Some(handle) = self.handle.take() {
-            // Replace our reciever with dummy one, so we can drop and close the
-            // one actually communicating with the thread
-            let recv = std::mem::replace(&mut self.message_recv, never());
-
-            // Dropping the original reciever initiates thread sub-process shutdown
-            drop(recv);
```

Comment:
This is unrelated, but interesting. 

This bit of code rubs me the wrong way, because we base cancellation on dropping the receiver, and not sender.  This is violates the unidirectional sender->receiver information flow, and is considered "bad".

It took me a while to realize what specific problem we have in this case, but I think there *is* a problem (mostly theoretical, which doesn't need fixing). Consider a case where cargo is slow at producing messages, and where we are waiting for the next message from Cargo. In this situation, even if we drop the receiver, we won't notice the cancellation until we get *next* message from Cargo (which might take a while). The "proper" way to do this would be to store a (dummy) reciever in the thread, and do a `selec!` on the event of receiving cargo message and on the event of dropped sender. That way, we notice cancellation immediately and can cancel Cargo without blocking.

Reply:
This reminds me of that dummy receiver being the cancellation token and its appropriate sender the cancellation token source.

Reply:
Nice observation, seems a reasonable thing to change at some point.